### ISSUE TITLE
Keras Model Conversion

### ISSUE THREAD

**Issue by formulajuan** [sassoftware/python-dlpy#260]
Hi, I have trained a model in Keras and saved the weights/architecture to a h5 file and I want to convert it to a DLPy model to deploy it/make inference. I'm able to convert the model but I can't seem to attach the SAS compatible h5 weights file that is generated automatically from the `Model.from_keras_model` function. My code and the associated error message is below:

```
from keras.models import load_model
from dlpy import Model

keras_resnet = load_model("/images/h5models/Digit_Classifier_v2.h5")

model3, use_gpu = Model.from_keras_model(s, 
                                         keras_model=keras_resnet, 
                                         output_model_table='Wide-ResNet-Classifier',
                                         include_weights=True,
                                         offsets=mean,
                                         std=std,
                                         verbose=False)
```
NOTE: Model table is attached successfully!
NOTE: Model is named to "wide-resnet-classifier" according to the model name in the table.
NOTE: the model weights has been stored in the following file:
/home/viyademo01/Wide-ResNet-Classifier_weights.kerasmodel.h5

`model3.load_weights_from_keras(path="/home/viyademo01/Wide-ResNet-Classifier_weights.kerasmodel.h5")`

![image](https://user-images.githubusercontent.com/57802244/70430580-04a98080-1a73-11ea-8196-a330e68aabd0.png)

The error message suggests moving it to a different location but the viyademo01 user only has write permissions to that directory ("/home/viyademo01"). I'm wondering if this is something I can fix in code or if it is permissions based? I've tried this using both "Public" and "CASUSER(viyademo01)" as the active caslib but both produce the same problem since it tries to create a new caslib. 

The system output also notes that the model weights have been attached successfully which I don't think is correct.

**Comment by formulajuan** [comment]
I fixed this by moving the SAS compatible h5 file to a different directory and loading from there.

**Comment by sasxianhu** [comment]
Here is why it happens and your solution is indeed the fastest one. We try to create a new caslib based on your h5 path by first checking whether the h5 path is already used as one of caslibs or not. If yes, we use yours. If not, we create one. Obviously, our check somehow is wrong in your case

**Comment by sasxianhu** [comment]
BTW: do you use DLPy by git clone or pip install?

**Comment by formulajuan** [comment]
> BTW: do you use DLPy by git clone or pip install?

We were provided with a pre-configured environment, I'm not sure.

**Comment by formulajuan** [comment]
I have an unrelated question (sorry), I'm able to load and deploy my keras model using DLPy but I'm unable to use Model.predict the same way as the keras model.predict function.  

For example, in keras I prepare the image for classification like this:

```
# resize
image = cv2.resize(image, (32,32))

# expand dims from [32,32,3] to [1,32,32,3]
image = np.expand_dims(image, axis=0)

# standardisation
image = (image - mean) / std

# make prediction
pred = model.predict(image)
```

And an array of 10 possible classes are returned

```
array([[9.9963045e-01, 3.4759854e-05, 2.2671124e-05, 1.0784574e-05,
        2.7091713e-05, 2.5261865e-05, 1.0057416e-04, 4.0211271e-05,
        3.3698216e-05, 7.4550153e-05]], dtype=float32)
```

I tried passing my mean and std values using the DLPy Model.from_keras_model (as in my original post) and resizing the images for testing using the ImageTable.resize function but there's an error when calling Model.predict:

![image](https://user-images.githubusercontent.com/57802244/70452438-04bf7580-1a9f-11ea-8f9f-b33052159c44.png)

I've tried not providing any mean/std data and using the default ImageNet channel means with no success. 

I'm not sure what pre-processing the Model.predict function carries out by default for an image classification model, or if my DLPy model should actually be an image classification model in this case?

**Comment by wenyushi** [comment]
The error will happen when BN layer contains small variances( < 10e-6). I will fix the issue in my next PR. BTW, one workaround is to convert the model into ONNX and then load the model from ONNX.

**Comment by formulajuan** [comment]
Just to clarify, do you mean using model.deploy (with output_format='onnx') and loading from CAS after it has been deployed? Or just directly loading a local ONNX file using Model.from_onnx_model?

**Comment by wenyushi** [comment]
I mean loading a local ONNX file using Model.from_onnx_model.

**Comment by formulajuan** [comment]
Okay thanks. I'm not sure if there is a problem with my model then because the onnx_to_sas function that is called during the Model.from_onnx_model function throws an error:

![image](https://user-images.githubusercontent.com/57802244/70543903-15d1ba80-1b63-11ea-883f-207942d6e722.png)

For reference, I converted my existing keras h5 model to onnx using [this example.](https://github.com/onnx/onnx-docker/blob/master/onnx-ecosystem/converter_scripts/keras_onnx.ipynb)

**Comment by wenyushi** [comment]
SAS ONNX parser doesn't support [transpose ](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Transpose)op. Consider it is a resnet model, I believe the operation happens at input layer used for reordering axis, eg. converting BHWC to BCHW. If so, we can skip the operation. @yg2288

**Comment by yg2288** [comment]
In some cases it's possible to skip the transpose after input.  However if there's a fully connected layer in the model (I assume in this case with resnet), typically onnx converters will insert a transpose before that too, so the FC weights will need to be reordered by DLPy.  I can work on adding this support, will open an issue to track this.  Thanks

**Comment by yg2288** [comment]
Hi @formulajuan I'm trying to recreate the "Wide-ResNet-Classifier" from Keras and use onnxmltools to convert it.  Could you point to the code to build this Keras model?  Thanks.

**Comment by formulajuan** [comment]
Hi, I used [this code](https://github.com/titu1994/Wide-Residual-Networks/blob/master/wide_residual_network.py) to build my Keras model. I can send you my h5 file as well if that helps.

**Comment by yg2288** [comment]
Yes, that would be helpful, thanks!

**Comment by formulajuan** [comment]
Here you go, please let me know if there's anything else I can help with.
[ResNet_Models.zip](https://github.com/sassoftware/python-dlpy/files/3978938/ResNet_Models.zip)

**Comment by yg2288** [comment]
Hi @formulajuan 
Please see the attached notebook for importing this model.  This requires DLPy '1.1.3-dev' and ONNX 1.6.  Let me know if you have any questions.
Thanks!

[onnx_notebook.zip](https://github.com/sassoftware/python-dlpy/files/4059406/onnx_notebook.zip)

**Comment by formulajuan** [comment]
Hi @yg2288 I was able to convert the model and load it into CAS as expected, thanks for your help!

I do have a question about the pre-processing of images that goes on under the hood because my results using the converted model are similar but slightly worse than the original Keras model. As I said in an earlier comment I need to replicate the pre-processing used for the Keras model:

> `# resize`
> `image = cv2.resize(image, (32,32))`
> 
> `# standardisation`
> `image = (image - mean) / std`


I noticed that the [Model.from_keras_model](https://sassoftware.github.io/python-dlpy/generated/dlpy.model.Model.from_keras_model.html) and [Model.from_onnx_model](https://sassoftware.github.io/python-dlpy/generated/dlpy.model.Model.from_onnx_model.html) functions include optional arguments for both offsets (mean) and std. The from_keras function has the option of providing both arguments as pre-defined lists but the from_onnx function only supports the offsets as a pre-defined list and an std value of midrange, std, or none. Is there a reason that these are different between the two functions?

The closest I got to replicating my original results was by carrying out the pre-processing on the images separately before placing them in an ImageTable. I compared that to passing my list of offsets and the string "std" to the `from_onnx_model` function but the results were mostly wrong. I was wondering if the results would be better if I could pass both the mean and std lists that I use in Keras? Thanks.

**Comment by yg2288** [comment]
Hi @formulajuan , 
Recently `norm_stds` was added as a parameter in `from_onnx_model`, used like this:
`norm_stds=(255*0.229, 255*0.224, 255*0.225), offsets=(255*0.485, 255*0.456, 255*0.406)`.   

I see this wasn't present in the 1.1.2 release - could you try upgrading to master using git+https://github.com/sassoftware/python-dlpy.git@master

Thanks

**Comment by formulajuan** [comment]
Yeah the `norm_stds` parameter works much better. The classifications are now correct for a small sample and the confidences are more aligned to the equivalent Keras values but some vary more than others (e.g. 0.75 confidence vs 0.99). 

I'm curious to see if there's any difference between using cv2.resize and ImageTable.resize to account for the slight variation in confidence scores.

**Comment by yg2288** [comment]
Thanks for reporting the difference in output, I am looking into this.

**Comment by sasxianhu** [comment]
Let me close this one. For the resize issue, feel free to open a new one to track it.
